airflow/providers/google/cloud/operators/bigquery.py

```diff
@@ -2039,6 +2039,155 @@ def execute(self, context) -> None:
+    )
+    template_fields_renderers = {"schema_fields": "json"}
+    ui_color = BigQueryUIColors.TABLE.value
+
+    @classmethod
+    def _patch_schema(cls, old_schema: Dict, new_schema: Dict) -> None:
+        """
+        Updates the content of "old_schema" with
+        the fields from new_schema. Makes changes
+        in place and hence has no return value.
+        Works recursively for sub-records.
+        Start by turning the schema list of fields into
+        a dict keyed on the field name for both the old
+        and the new schema.
+
+        :param old_schema: Old schema which is updated in-place
+        :type old_schema: dict
+        :param new_schema: Partial schema definition used to patch old schema
+        :type new_schema: dict
+        """
+        new_fields = {field["name"]: field for field in new_schema["fields"] if "name" in field}
+        old_fields = {field["name"]: field for field in old_schema["fields"]}
+
+        # Iterate over all new fields and update the
+        # old_schema dict
+        for field_key in new_fields.keys():
+            # Check if the field exists in the old_schema, if
+            # so change it
+            if field_key in old_fields:
+                old_field = old_fields[field_key]
+                new_field = new_fields[field_key]
+                # Check if recursion is needed
+                if "fields" in new_field:
+                    cls._patch_schema(old_field, new_field)
+                    del new_field["fields"]
+
+                # Do the update
+                old_field.update(new_field)
+
+            # Field didn't exist, add it as a new field
+            else:
+                old_schema["fields"].append(new_fields[field_key])
+
+    @apply_defaults
+    def __init__(
+        self,
+        *,
+        dataset_id: Optional[str] = None,
+        schema_fields: List[Dict[str, Any]],
+        table_id: Optional[str] = None,
+        project_id: Optional[str] = None,
+        gcp_conn_id: str = 'google_cloud_default',
+        delegate_to: Optional[str] = None,
+        impersonation_chain: Optional[Union[str, Sequence[str]]] = None,
+        **kwargs,
+    ) -> None:
+        self.schema_fields = schema_fields
+        self.table_id = table_id
+        self.dataset_id = dataset_id
+        self.project_id = project_id
+        self.gcp_conn_id = gcp_conn_id
+        self.delegate_to = delegate_to
+        self.impersonation_chain = impersonation_chain
+        super().__init__(**kwargs)
+
+    def execute(self, context):
+        bq_hook = BigQueryHook(
+            gcp_conn_id=self.gcp_conn_id,
+            delegate_to=self.delegate_to,
+            impersonation_chain=self.impersonation_chain,
+        )
+
+        table_schema = bq_hook.get_schema(
+            dataset_id=self.dataset_id, table_id=self.table_id, project_id=self.project_id
+        )
+
+        self._patch_schema(old_schema=table_schema, new_schema={"fields": self.schema_fields})
```

Comment:
in-place modification of the variable can very tricky, why not create a new variable `updated_schema`?

Reply:
In order to introduce that variable I'd also need to introduce some version of deepcopy . I guess if I move it into a hook instead that's already in there.

Reply:
Not sure I agree that it is tricky in a setting like this though, it's just a mutable object that I change